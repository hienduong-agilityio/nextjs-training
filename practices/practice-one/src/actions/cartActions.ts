'use server';

// Libraries
import { revalidatePath } from 'next/cache';

// Services
import { addToCart, deleteFromCart } from '@/services';

// Constants
import { ROUTE, TOAST_MESSAGES } from '@/constants';

/**
 * Server action to add a product to the cart.
 * @param userId - The ID of the user
 * @param productId - The ID of the product
 * @param quantity - Quantity to add
 * @param maxQuantity - Max quantity to add
 * @returns {boolean} - Success or failure of the operation
 */
export async function handleAddToCart({
  userId,
  productId,
  quantity,
  maxQuantity,
}: {
  userId: number;
  productId: string;
  quantity: number;
  maxQuantity: number;
}): Promise<boolean> {
  try {
    const { success } = await addToCart(userId, {
      productId,
      quantity,
      maxQuantity,
    });

    if (success) {
      revalidatePath(ROUTE.CART);
      revalidatePath(ROUTE.ROOT);
    }

    return success;
  } catch (error) {
    throw new Error(TOAST_MESSAGES.API_ERROR);
  }
}

/**
 * Server action to remove a product from the cart.
 * @param userId - The ID of the user
 * @param productId - The ID of the product to remove
 * @returns {boolean} - Success or failure of the operation
 */
export async function handleRemoveFromCart({
  userId,
  productId,
}: {
  userId: number;
  productId: string;
}): Promise<boolean> {
  try {
    const { success } = await deleteFromCart(userId, productId);

    if (success) {
      revalidatePath(ROUTE.CART);
      revalidatePath(ROUTE.ROOT);
    }

    return success;
  } catch (error) {
    throw new Error(TOAST_MESSAGES.API_ERROR);
  }
}
