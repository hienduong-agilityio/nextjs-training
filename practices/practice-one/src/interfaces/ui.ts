export interface IInputGroupProps {
  placeholder?: string;
  value?: string;
  buttonText?: string;
  inputName?: string;
  isDisabled?: boolean;
  isLoading?: boolean;
  startIcon?: React.ReactNode;
  customClass?: {
    container?: string;
    inputContainer?: string;
    input?: string;
    button?: string;
  };
  onSubmit?: (event: React.FormEvent<HTMLFormElement>) => void;
}
