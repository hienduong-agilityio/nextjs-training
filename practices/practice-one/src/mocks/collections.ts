import { Category } from '@/types/category';

type HotDeal = {
  name: Category | 'all';
  count: number;
};

export const HOT_DEALS: HotDeal[] = [
  { name: 'all', count: 2 },
  { name: 'beauty', count: 2 },
  { name: 'fragrances', count: 48 },
  { name: 'furniture', count: 95 },
  { name: 'groceries', count: 23 },
];

export const BRANDS = [
  { name: 'Nike', count: 99 },
  { name: 'Adidas', count: 99 },
  { name: 'Siemens', count: 99 },
];

export const FILTER_GROUP = [{ title: 'Hot Deals', items: HOT_DEALS }];
