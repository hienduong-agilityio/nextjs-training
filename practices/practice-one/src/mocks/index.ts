export * from './product';
export * from './service';
export * from './footer';
export * from './collections';
