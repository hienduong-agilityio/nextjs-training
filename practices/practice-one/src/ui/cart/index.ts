export * from './CartTable';
export * from './OrderSummary';
export * from './VoucherInput';
export * from './EmptyCart';

export { default as CartItemRow } from './CartItemRow';
