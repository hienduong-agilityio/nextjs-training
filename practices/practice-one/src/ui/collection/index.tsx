export * from './LoadingCategoryGroup';
export * from './LoadingFilterSortBar';
