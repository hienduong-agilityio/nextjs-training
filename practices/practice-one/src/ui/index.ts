export * from './products/index';
export * from './collection/index';
