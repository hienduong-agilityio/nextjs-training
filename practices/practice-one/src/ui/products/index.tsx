export * from './CartAndFavoriteActions';
export * from './ProductImages';
export * from './ProductInfo';
export * from './QuantityControl';
export * from './SocialShare';
export * from './RelatedProduct';
export * from './ProductDetailTabs';
export * from './ProductReview';
export * from './ProductActions';
export * from './ProductLoading';
