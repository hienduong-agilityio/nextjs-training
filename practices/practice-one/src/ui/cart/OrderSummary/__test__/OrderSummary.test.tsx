import { render, screen, fireEvent, act } from '@testing-library/react';
import { useRouter } from 'next/navigation';

// Actions
import { handleClearProductFromCart } from '@/actions';

// UI
import { OrderSummary } from '@/ui';

// Mocks
import { ORDER_SUMMARY } from '@/mocks';

// Mock dependencies
jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
}));

jest.mock('@/actions', () => ({
  handleClearProductFromCart: jest.fn(),
}));

jest.mock('react', () => {
  const actualReact = jest.requireActual('react');
  return {
    ...actualReact,
    useOptimistic: jest.fn((initialState) => {
      let state = initialState;
      const setState = jest.fn((newValues) => {
        state = { ...state, ...newValues };
      });
      return [state, setState];
    }),
  };
});

describe('OrderSummary Component', () => {
  const mockRouterPush = jest.fn();

  const renderOrderSummary = (summary = ORDER_SUMMARY.withData) =>
    render(<OrderSummary summary={summary} />);

  beforeEach(() => {
    jest.clearAllMocks();
    (useRouter as jest.Mock).mockReturnValue({ push: mockRouterPush });
  });

  it('matches snapshot', () => {
    const { asFragment } = renderOrderSummary();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly with default props', () => {
    renderOrderSummary();

    expect(screen.getByText('Subtotal')).toBeInTheDocument();
    expect(screen.getByText('$100.00')).toBeInTheDocument();
    expect(screen.getByText('Shipping fee')).toBeInTheDocument();
    expect(screen.getByText('$10')).toBeInTheDocument();
    expect(screen.getByText('Coupon')).toBeInTheDocument();
    expect(screen.getByText('$20.00')).toBeInTheDocument();
    expect(screen.getByText('Total')).toBeInTheDocument();
    expect(screen.getByText('$90.00')).toBeInTheDocument();
  });

  it('disables the "Check out" button when subtotal is zero', () => {
    renderOrderSummary(ORDER_SUMMARY.emptyData);

    const button = screen.getByRole('button', { name: /Check out/i });
    expect(button).toBeDisabled();
  });

  it.each([true, false])(
    'calls handleClearProductFromCart and navigates to correct status on checkout (success: %s)',
    async (actionResolvedValue) => {
      const expectedStatus = actionResolvedValue ? 'success' : 'error';
      (handleClearProductFromCart as jest.Mock).mockResolvedValue(
        actionResolvedValue,
      );

      renderOrderSummary();
      const button = screen.getByRole('button', { name: /Check out/i });

      await act(async () => {
        fireEvent.click(button);
      });

      expect(handleClearProductFromCart).toHaveBeenCalledWith(
        expect.any(Number),
      );
      expect(mockRouterPush).toHaveBeenCalledWith(
        expect.stringContaining(`status=${expectedStatus}`),
      );
    },
  );
});
