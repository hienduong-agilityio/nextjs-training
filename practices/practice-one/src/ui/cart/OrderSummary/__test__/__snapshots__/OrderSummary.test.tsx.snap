// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderSummary Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-full lg:w-1/2 2xl:w-1/4 space-y-4 p-4"
  >
    <div
      class="flex flex-col gap-4 border-b-2 pb-4 border-secondary-100"
    >
      <div
        class="flex justify-between text-sm md:text-base lg:text-lg"
      >
        <span>
          Subtotal
        </span>
        <span>
          $100.00
        </span>
      </div>
      <div
        class="flex justify-between text-sm md:text-base lg:text-lg"
      >
        <span>
          Shipping fee
        </span>
        <span>
          $10
        </span>
      </div>
      <div
        class="flex justify-between text-sm md:text-base lg:text-lg"
      >
        <span>
          Coupon
        </span>
        <span>
          $20.00
        </span>
      </div>
    </div>
    <div
      class="flex justify-between text-lg md:text-2xl lg:text-3xl py-4"
    >
      <span>
        Total
      </span>
      <span>
        $90.00
      </span>
    </div>
    <button
      class="flex items-center justify-center gap-2 rounded p-2 disabled:cursor-not-allowed disabled:opacity-50 bg-opacity-100 shadow-md bg-primary-200 text-white disabled:hover:bg-primary-200 hover:bg-primary-400 w-full py-3 md:py-4 text-sm md:text-base lg:text-lg"
      type="button"
    >
      Check out
    </button>
  </div>
</DocumentFragment>
`;
